### PR TITLE
Better ambiguous column detection

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description (up to few sentences):
Detect and throw if we cannot detect table for column name in query.

### CODE REVIEW

File: dbms/src/Interpreters/SyntaxAnalyzer.cpp [hunk 1]
```diff
@@ -541,11 +565,16 @@ void collectJoinedColumns(AnalyzedJoin & analyzed_join, const ASTSelectQuery & s
     }
 }
 
-void replaceJoinedTable(const ASTTablesInSelectQueryElement* join)
+void replaceJoinedTable(const ASTTablesInSelectQueryElement * join)
 {
     if (!join || !join->table_expression)
         return;
 
+    /// TODO: Push down for CROSS JOIN is not OK [disabled]
```
Comment (4ertus2) [inline dbms/src/Interpreters/SyntaxAnalyzer.cpp]: I've also disabled table to subquery rewrite logic for CROSS JOINs here cause it doesn't work properly.

**Comment by akuzm** [comment]
`00094_union_race_conditions_5` fails, is this relevant?

File: dbms/tests/queries/0_stateless/01018_anbiguous_column.reference [hunk 2]
```diff
@@ -0,0 +1,12 @@
+0	0
```
Comment (akuzm) [inline dbms/tests/queries/0_stateless/01018_anbiguous_column.reference]: The file name has a typo -- aNbigous.

File: dbms/src/Interpreters/IdentifierSemantic.cpp [hunk 3]
```diff
@@ -5,6 +5,61 @@
...
+    best_table_pos = 0;
+    auto best_match = ColumnMatch::NoMatch;
+    size_t same_match = 0;
+
+    for (size_t i = 0; i < tables.size(); ++i)
+    {
+        auto match = IdentifierSemantic::canReferColumnToTable(identifier, extractTable(tables[i]));
+        if (value(match))
+        {
+            if (value(match) > value(best_match))
```
Comment (akuzm) [inline dbms/src/Interpreters/IdentifierSemantic.cpp]: enum values are comparable with `>`, we can omit the cast.

File: dbms/src/Interpreters/IdentifierSemantic.cpp [hunk 4]
```diff
@@ -5,6 +5,61 @@
...
+    using ColumnMatch = IdentifierSemantic::ColumnMatch;
+
+    best_table_pos = 0;
+    auto best_match = ColumnMatch::NoMatch;
+    size_t same_match = 0;
+
+    for (size_t i = 0; i < tables.size(); ++i)
+    {
+        auto match = IdentifierSemantic::canReferColumnToTable(identifier, extractTable(tables[i]));
+        if (value(match))
```
Comment (akuzm) [inline dbms/src/Interpreters/IdentifierSemantic.cpp]: So NoMatch is implicitly assigned 0, then we static_cast it to UInt32 using a special function, and then implicitly convert to bool? I'd just write `match != NoMatch`.

**Review by akuzm — APPROVED** [review]

**Pull request merged**